### PR TITLE
Add mongodb_auth role

### PR DESCRIPTION
**Pull request by cognifloyd**
##### SUMMARY
Add a new role to enable authentication on a random mongodb server.

Should handle:
- a server that does not have authentication enabled,
- a server that has auth enabled (or started with --auth) and localhost exception still applies, and
- a server that is already properly configured with auth enabled.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
mongodb_auth role

### CODE REVIEW

**Comment by cognifloyd** [comment]
I used this previously with mongo 3.4, but I'm having issues getting the checks to work with 4.2.
At least, I assume that it's the mongo version that's tripping me up.

With 3.4 and either auth is disabled, or localhost exception is active, the following command succeeds:
```shell
mongo admin --port 27017 --eval "db.getUsers()"
```

Now, it appears that the localhost exception is somehow not enabled for these installs in docker containers.
Any recommendations on how to better handle this?

File: roles/mongodb_auth/molecule/default/playbook.yml [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    - name: Check mongo auth
+      raw: 'mongo admin --port 27017 --eval "db.getUsers()"'
+      # raw: 'mongo admin --port 27017 --username admin --password admin --eval "db.getUsers()"'
+      changed_when: no
+      failed_when: no
+
+    - include_role:
+        name: mongodb_mongod
+      vars:
+      # does bindIp=0.0.0.0 remove localhost exception?
```
Comment (rhysmeister) [inline roles/mongodb_auth/molecule/default/playbook.yml:26]: This doesn't remove the localhost exception. I use it here... https://github.com/rhysmeister/AutomatingMongoDBWithAnsible

mongodb_mongod role default bindIp to 0.0.0.0

File: roles/mongodb_auth/molecule/default/tests/test_default.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+                           False,
+                           False)
+    return ansible
+
+
+def test_mongod_cnf_file(host):
+    f = host.file('/etc/mongod.conf')
+
+    assert f.exists
+    assert f["security"]["authorization"] == "enabled"
```
Comment (rhysmeister) [inline roles/mongodb_auth/molecule/default/tests/test_default.py:22]: Cool! Didn't know you could do this.

**Comment by rhysmeister** [comment]
> I used this previously with mongo 3.4, but I'm having issues getting the checks to work with 4.2.
> At least, I assume that it's the mongo version that's tripping me up.
> 
> With 3.4 and either auth is disabled, or localhost exception is active, the following command succeeds:
> 
> ```shell
> mongo admin --port 27017 --eval "db.getUsers()"
> ```
> 
> Now, it appears that the localhost exception is somehow not enabled for these installs in docker containers.
> Any recommendations on how to better handle this?



> ```shell
> mongo admin --port 27017 --eval "db.getUsers()"
> ```

@cognifloyd I think you probably can't run this command anymore - "Connections using the localhost exception only have access to create the first user on the admin database."

https://docs.mongodb.com/manual/core/security-users/

3.4 is pretty old so I'd only worry about supporting later versions.

File: roles/mongodb_auth/molecule/default/tests/test_default.py (hunk 2 above)
Comment (cognifloyd) [reply roles/mongodb_auth/molecule/default/tests/test_default.py:22]: I don't know if we can... I was guessing. If we can't then I'll have to open the file to check.

Comment (rhysmeister) [reply roles/mongodb_auth/molecule/default/tests/test_default.py:22]: OK. Probably not then. You'll need to use contains() or check the content_string var.

**Comment by cognifloyd** [comment]
OK, I redid this to be a bit closer to how you did it in the mongodb_mongos molecule playbook.

Then, the key issue was a problem with the `update_password` parameter. I misread the code - it has to be "always" to use the localhost exception, not "on_create".

So, now it seems to be working. :) I'll reenable all the travis tests once this run finishes to make sure I caught everything.

**Comment by cognifloyd** [comment]
Please cancel all but the latest Travis jobs.
You can see here that the new mongodb_auth role tests are passing:
https://travis-ci.com/github/cognifloyd/community.mongodb/builds/184829161

What do you think of this role?

**Comment by cognifloyd** [comment]
I'm waiting to merge #182 before cleaning up this PR.

File: .github/workflows/ansible-test.yml [hunk 3]
```diff
@@ -117,10 +117,10 @@ jobs:
           python-version: ${{ matrix.python_version }}
 
       - name: Install community.general
-        run: ansible-galaxy collection install community.general -p ../../
+        run: "for i in {1..3}; do ansible-galaxy collection install community.general -p ../../ && break || sleep 5; done"
```
Comment (rhysmeister) [inline .github/workflows/ansible-test.yml:120]: I think we should get rid of the retry loops here. Better solved at galaxy end and caching.

File: .github/workflows/ansible-test.yml [hunk 4]
```diff
@@ -117,10 +117,10 @@ jobs:
...
 
       - name: Install community.crypto
-        run: ansible-galaxy collection install community.crypto -p ../../
+        run: "for i in {1..3}; do ansible-galaxy collection install community.crypto -p ../../ && break || sleep 5; done"
```
Comment (rhysmeister) [inline .github/workflows/ansible-test.yml:123]: Ditto.

File: roles/mongodb_auth/tasks/main.yml [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+  package:
+    name:
+      - "{{ mongod_package }}"
+      # pyyaml is used to validate yaml files on change
+      - "{{ pyyaml_package }}"
+  register: _pkg
+  until: _pkg is succeeded
+  retries: 5
+
+- name: Warn about default credentials
```
Comment (rhysmeister) [inline roles/mongodb_auth/tasks/main.yml:27]: Nice touch!

**Review by rhysmeister — COMMENTED** [review]
Top work

File: .github/workflows/ansible-test.yml (hunk 3 above)
Comment (cognifloyd) [reply .github/workflows/ansible-test.yml:120]: Oops. I didn't mean to add these here. removing...

**Comment by cognifloyd** [comment]
I resolved the issue you raised, and tests all pass, so I'm merging this.

**Pull request merged**